### ISSUE TITLE
Copying CXX header __string/char_traits.h FAILED (Compiling LLVM)

### ISSUE THREAD

**Issue by AA1999** [llvm/llvm-project#56865]
I have done a git pull on my current llvm-project clone and to configure it for installation I did this command:

```shell
cmake -S llvm -B build -G Ninja -DLLVM_ENABLE_PROJECTS='clang;clang-tools-extra;cross-project-tests;flang;libc;libclc;lld;lldb;openmp;polly' -DLLVM_ENABLE_RUNTIMES='libcxx;libcxxabi;compiler-rt;' -DLLVM_BUILD_DOCS=OFF -DLLVM_ENABLE_BINDINGS=OFF -DLIBCXX_CXX_ABI=libcxxabi
```

However upon compiling I get the following warnings and error

```console
CMake Warning at /home/arshia/Desktop/llvm-project/compiler-rt/cmake/Modules/CompilerRTUtils.cmake:353 (message):
  llvm-config finding testingsupport failed with status 1
Call Stack (most recent call first):
  /home/arshia/Desktop/llvm-project/compiler-rt/CMakeLists.txt:71 (load_llvm_config)


-- LLVM_MAIN_SRC_DIR: "/home/arshia/Desktop/llvm-project/runtimes/../llvm"
-- Compiler-RT supported architectures: x86_64;i386
-- Supported architectures for crt: i386;x86_64
CMake Warning (dev) at /usr/share/cmake/Modules/ExternalProject.cmake:2148 (message):
  Policy CMP0114 is not set: ExternalProject step targets fully adopt their
  steps.  Run "cmake --help-policy CMP0114" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  ExternalProject target 'libcxx_msan_x86_64' would depend on the targets for
  step(s) 'clean' under policy CMP0114, but this is being left out for
  compatibility since the policy is not set.
Call Stack (most recent call first):
  /home/arshia/Desktop/llvm-project/compiler-rt/cmake/Modules/AddCompilerRT.cmake:721 (ExternalProject_Add_StepTargets)
  /home/arshia/Desktop/llvm-project/compiler-rt/lib/msan/tests/CMakeLists.txt:140 (add_custom_libcxx)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/share/cmake/Modules/ExternalProject.cmake:2148 (message):
  Policy CMP0114 is not set: ExternalProject step targets fully adopt their
  steps.  Run "cmake --help-policy CMP0114" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  ExternalProject target 'libcxx_tsan_x86_64' would depend on the targets for
  step(s) 'clean' under policy CMP0114, but this is being left out for
  compatibility since the policy is not set.
Call Stack (most recent call first):
  /home/arshia/Desktop/llvm-project/compiler-rt/cmake/Modules/AddCompilerRT.cmake:721 (ExternalProject_Add_StepTargets)
  /home/arshia/Desktop/llvm-project/compiler-rt/lib/tsan/CMakeLists.txt:35 (add_custom_libcxx)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/share/cmake/Modules/ExternalProject.cmake:2148 (message):
  Policy CMP0114 is not set: ExternalProject step targets fully adopt their
  steps.  Run "cmake --help-policy CMP0114" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  ExternalProject target 'libcxx_fuzzer_i386' would depend on the targets for
  step(s) 'clean' under policy CMP0114, but this is being left out for
  compatibility since the policy is not set.
Call Stack (most recent call first):
  /home/arshia/Desktop/llvm-project/compiler-rt/cmake/Modules/AddCompilerRT.cmake:721 (ExternalProject_Add_StepTargets)
  /home/arshia/Desktop/llvm-project/compiler-rt/lib/fuzzer/CMakeLists.txt:161 (add_custom_libcxx)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/share/cmake/Modules/ExternalProject.cmake:2148 (message):
  Policy CMP0114 is not set: ExternalProject step targets fully adopt their
  steps.  Run "cmake --help-policy CMP0114" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  ExternalProject target 'libcxx_fuzzer_x86_64' would depend on the targets
  for step(s) 'clean' under policy CMP0114, but this is being left out for
  compatibility since the policy is not set.
Call Stack (most recent call first):
  /home/arshia/Desktop/llvm-project/compiler-rt/cmake/Modules/AddCompilerRT.cmake:721 (ExternalProject_Add_StepTargets)
  /home/arshia/Desktop/llvm-project/compiler-rt/lib/fuzzer/CMakeLists.txt:161 (add_custom_libcxx)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generated Sanitizer SUPPORTED_TOOLS list on "Linux" is "asan;lsan;msan;tsan;ubsan"
-- sanitizer_common tests on "Linux" will run against "asan;lsan;msan;tsan;ubsan"
-- Supported architectures for crt: i386;x86_64
-- check-shadowcallstack does nothing.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins
[54/73] Performing build step for 'runtimes'
[5/1565] Copying CXX header __string/char_traits.h
FAILED: /home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/char_traits.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h /home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/char_traits.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h" to "/home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/char_traits.h".
[6/1565] Copying CXX header __string/extern_template_lists.h
FAILED: /home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/extern_template_lists.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h /home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/extern_template_lists.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h" to "/home/arshia/Desktop/llvm-project/build/include/c++/v1/__string/extern_template_lists.h".
[14/1565] Copying CXX header __type_traits/aligned_storage.h
ninja: build stopped: subcommand failed.
FAILED: runtimes/runtimes-stamps/runtimes-build /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-stamps/runtimes-build
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins && /usr/bin/cmake --build .
[58/73] Building CXX object third-party/benchmark/src/CMakeFiles/benchmark.dir/benchmark_register.cc.o
ninja: build stopped: subcommand failed.
```

My os is Arch-Linux and my current clang compiling this is 15.0.0

Exact details:
```console
➜  ~ clang --version
clang version 15.0.0 (https://github.com/llvm/llvm-project.git f6112f490cfeed05b42ae33cc994292000aee0da)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin
```

**Comment by philnik777** [comment]
You have to remove `include/c++/v1/__string` from your build directory.

**Comment by AA1999** [comment]
Will it not result in any missing files from the final build?

**Comment by AA1999** [comment]
```cmake
[246/523] Copying CXX header __string/char_traits.h
FAILED: include/c++/v1/__string/char_traits.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/char_traits.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/char_traits.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h" to "/home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/char_traits.h".
[247/523] Copying CXX header __string/extern_template_lists.h
FAILED: include/c++/v1/__string/extern_template_lists.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/extern_template_lists.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/extern_template_lists.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h" to "/home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string/extern_template_lists.h".
[255/523] Copying CXX header __support/xlocale/__posix_l_fallback.h
ninja: build stopped: subcommand failed.
[1473/1553] Building CXX object c...rc.x86_64.dir/dlfcn_wrapper.cpp.o
FAILED: compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/src/libcxx_fuzzer_i386-stamp/libcxx_fuzzer_i386-build compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/lib/libc++.a compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/lib/libc++abi.a /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/src/libcxx_fuzzer_i386-stamp/libcxx_fuzzer_i386-build /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/lib/libc++.a /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/lib/libc++abi.a
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386 && /usr/bin/cmake --build .
[1483/1553] Building CXX object c...est/googletest/src/gtest-all.cc.o
ninja: build stopped: subcommand failed.
[53/56] Linking CXX static library lib/libbenchmark_main.a
FAILED: runtimes/runtimes-stamps/runtimes-build /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-stamps/runtimes-build
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins && /usr/bin/cmake --build .
ninja: build stopped: subcommand failed.
```
Even after deleting the folder.

**Comment by philnik777** [comment]
It looks like you also have to delete `runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_i386/include/c++/v1/__string`. You don't have to clear the build directory every time. The problem is just that the file `__string` has been replaced with the directory `__string` and CMake doesn't remove the file. You can of course clear the build directory every time, but I think in this case rebuilding everything takes a lot longer than removing e few files.

**Comment by AA1999** [comment]
```cmake
FAILED: include/c++/v1/__string/char_traits.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/char_traits.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/char_traits.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/char_traits.h" to "/home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/char_traits.h".
[248/523] Copying CXX header __string/extern_template_lists.h
FAILED: include/c++/v1/__string/extern_template_lists.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/extern_template_lists.h
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/libcxx/include && /usr/bin/cmake -E copy_if_different /home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/extern_template_lists.h
Error copying file (if different) from "/home/arshia/Desktop/llvm-project/libcxx/include/__string/extern_template_lists.h" to "/home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string/extern_template_lists.h".
[254/523] Copying CXX header __support/xlocale/__nop_locale_mgmt.h
ninja: build stopped: subcommand failed.
[100/135] Building CXX object compiler-rt/lib/fuzzer/CMakeFiles/RTfuzzer_main.i386.dir/FuzzerMain.cpp.o
FAILED: compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/src/libcxx_fuzzer_x86_64-stamp/libcxx_fuzzer_x86_64-build compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++abi.a /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/src/libcxx_fuzzer_x86_64-stamp/libcxx_fuzzer_x86_64-build /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++abi.a
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64 && /usr/bin/cmake --build .
[103/135] Building CXX object compiler-rt/lib/fuzzer/CMakeFiles/RTfuzzer.i386.dir/FuzzerMerge.cpp.o
ninja: build stopped: subcommand failed.
[8236/8239] Linking CXX static library projects/openmp/libomptarget/libomptarget.devicertl.a
FAILED: runtimes/runtimes-stamps/runtimes-build /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-stamps/runtimes-build
cd /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins && /usr/bin/cmake --build .
ninja: build stopped: subcommand failed.
```

Still the same looks like

**Comment by philnik777** [comment]
Not exactly. This time it's `runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/include/c++/v1/__string`. compiler-rt is making a lot of copies. That should get fixed at some point.

**Comment by AA1999** [comment]
```cmake
0/1] Re-running CMake...
-- Could NOT find LLVM (missing: LLVM_DIR)
-- Could NOT find Clang (missing: Clang_DIR)
-- Using libc++abi testing configuration: /home/arshia/Desktop/llvm-project/libcxxabi/test/lit.site.cfg.in
-- Could not find ParallelSTL, libc++abi will not attempt to use it but the build may fail if the libc++ in use needs it to be available.
-- Using libc++ testing configuration: /home/arshia/Desktop/llvm-project/libcxx/test/configs/llvm-libc++-static.cfg.in
CMake Warning at /home/arshia/Desktop/llvm-project/libcxx/CMakeLists.txt:381 (message):
  LIBCXX_ENABLE_STATIC_ABI_LIBRARY is an experimental option


-- check-runtimes does nothing.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/arshia/Desktop/llvm-project/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64
[272/272] Linking CXX static library lib/libc++experimental.a
[182/182] Linking CXX static library /home/arshia/Desk.../x86_64-unknown-linux-gnu/libclang_rt.fuzzer_no_main.a
[1931/1956] No install step for 'runtimes'
[1947/1948] Generating ../../../../include/flang/ieee_arithmetic.mod
```

This should be good, right?

**Comment by philnik777** [comment]
Yes, looks good. To make sure, running `ninja` again should say `no work to do.`.
